packages/firestore/src/lite-api/types.ts

```diff
@@ -34,16 +34,29 @@ export type Primitive = string | number | boolean | undefined | null;
 export type NestedUpdateFields<T extends Record<string, unknown>> =
   UnionToIntersection<
     {
-      // Check that T[K] extends Record to only allow nesting for map values.
-      [K in keyof T & string]: T[K] extends Record<string, unknown>
-        ? // Recurse into the map and add the prefix in front of each key
-          // (e.g. Prefix 'bar.' to create: 'bar.baz' and 'bar.qux'.
-          AddPrefixToKeys<K, UpdateData<T[K]>>
-        : // TypedUpdateData is always a map of values.
-          never;
+      [K in keyof T & string]: ChildUpdateFields<K, T[K]>;
     }[keyof T & string] // Also include the generated prefix-string keys.
   >;
 
+/**
+ * Helper for calculating the nested fields for a given type T1. This is needed
+ * to distribute union types such as `undefined | {...}` (happens for optional
+ * props) or `{a: A} | {b: B}`.
+ *
+ * In this use case, `V` is used to distribute the union types of `T[K]` on
+ * `Record`, since `T[K]` is evaluated as an expression and not distributed.
+ *
+ * See https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types
+ */
+export type ChildUpdateFields<K extends string, V> =
```

Comment:
This looks identical to the initial implementation. What has changed  that adds the support for optional fields?

Reply:
This is an excerpt from the documentation link:
> Conditional types in which the checked type is a naked type parameter are called distributive conditional types. Distributive conditional types are automatically distributed over union types during instantiation. For example, an instantiation of T extends U ? X : Y with the type argument A | B | C for T is resolved as (A extends U ? X : Y) | (B extends U ? X : Y) | (C extends U ? X : Y).

Here's a TS playground [link](https://www.typescriptlang.org/play?#code/PTAEBUAsEsGdQC4E8AOBTUdQDsD22BaAEzgQCdoAjAVwTSNAENsGy0U3Y1sF4FIMaAB6lo2AOagABgDk0sOkQCqKIozoAxaGgA2RWFMSo0AOgBQydKA25cAHnChhdFvABKaAMa4yROwooJABpqbABrPAB3bAA+GNAAXlAAbzMAbQBpTGxQMLQkXAAzCFAAMlAAsXEAXQAuCEzqpyEXfVAPb19-ciqg0FCI3GiYs1Ax8bGAfgaM6tGJ8frsNAA3NDIzAF8AbjMzEAgYPmNMeGWFeiYWUBJKmjo+ARx5RVBC7T0+yJhPSFA2Djybi8RBPX7McQYAC2jCIGDEoIwAAU3BYTjZcAB1aD8TFkRgodBkBzNVruLw+PyVYLMJBxRIpeaZbK5fJFErlak1eoAEVIFHuaHAxgcjRGOzRVj5d1oQpFADV4kl5aTuG0OpTuoFxH0BlFYvMFqBpvLDQslqt1rt9mAVGo6Dz1IxQJEcX8xMsyKBcCgENB8IwdG8PkRzJYMHb1GhHQhGNj+AB5X3+7CBuQXIhaXR+JWMsbnRT1VITDg++pc0AAH36LDQ72WRHmmyrNbh9foW2tBygGAUzDUvkwUJQOjQUOB6hTThWgeoUfg4EaTHgzGagNgsBT9SkAApF7MW6E22J6ABKVWudoUroADSCAE0YoZA5FGEh4JxcDo1nxcNJlmsZBSDaFRiJ4GBSHkBTFDuR51ieDDVvu1SnoYWBSMkmxSOY3jYAoiCwAAomQZA+PUGJ2JGDpOvGkBJn6AY6OmihZnoubFs8GZFmapYoPUcHto2YybJ2ewHDuKpJMhh61oJ549jc-JULKDDQMOo7jjwk74NOs7zqAKqMCuikygx2B9Pi-DrKCq6rsI66bvg24qs4arkp0fh3o+wEHFgqx6a8RnbgJCEXuq16eQ+8TVjMTSuZeGq3lFUhfD87rwDOOjQKGZi4fheDEaRZDkbYtF4gSRKUaoUYxnGbr0SmaYvPQrFEOx8wFvQ3ElqRfGtvBDZNqJQA) that shows the before and after.

Reply:
Thanks for the example! Now it makes sense to me.